### PR TITLE
MSALException needs to be serializable [962]

### PR DESCRIPTION
**Pull request by MarkZuber**
https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/issues/962

### CODE REVIEW

File: src/Microsoft.Identity.Client/MsalException.cs [hunk 1]
```diff
@@ -1,122 +1,188 @@
...
+            JObject jobj = new JObject();
+            PopulateJson(jobj);
+            return jobj.ToString();
+        }
+
+        private static readonly Lazy<Dictionary<string, Type>> s_typeNameToType = new Lazy<Dictionary<string, Type>>(() =>
+        {
+            return new Dictionary<string, Type>
+            {
+                { typeof(MsalException).Name, typeof(MsalException) },
```
Comment (jmprieur) [inline src/Microsoft.Identity.Client/MsalException.cs]: What would happen if we or customers would derive a new type from MsalException?

File: src/Microsoft.Identity.Client/MsalException.cs [hunk 2]
```diff
@@ -1,122 +1,188 @@
...
+namespace Microsoft.Identity.Client
+{
+    /// <summary>
+    /// Base exception type thrown when an error occurs during token acquisition.
+    /// For more details, see https://aka.ms/msal-net-exceptions
+    /// </summary>
+    /// <remarks>Avoid throwing this exception. Instead throw the more specialized <see cref="MsalClientException"/>
+    /// or <see cref="MsalServiceException"/>
+    /// </remarks>
+    public class MsalException : Exception
```
Comment (henrik-me) [inline src/Microsoft.Identity.Client/MsalException.cs]: >MsalException [](start = 17, length = 13)

I'm not sure I understand why it's not enough to mark these as serializable?

File: src/Microsoft.Identity.Client/MsalException.cs [hunk 3]
```diff
@@ -1,122 +1,188 @@
...
+            {
+                string errorCode = JsonUtils.GetExistingOrEmptyString(jobj, ErrorCodeKey);
+                string errorMessage = JsonUtils.GetExistingOrEmptyString(jobj, ErrorDescriptionKey);
+
+                MsalException ex = Activator.CreateInstance(exceptionType, errorCode, errorMessage) as MsalException;
+                ex.PopulateObjectFromJson(jobj);
+                return ex;
+            }
+
+            throw new MsalException();
```
Comment (henrik-me) [inline src/Microsoft.Identity.Client/MsalException.cs]: >throw new [](start = 12, length = 9)

shouldn't this return ? Alternatively the method should be ThrowExceptionFromJsonString()

File: src/Microsoft.Identity.Client/MsalException.cs (hunk 1 above)
Comment (henrik-me) [reply src/Microsoft.Identity.Client/MsalException.cs]: We wouldn't find it here. However not sure I understand the usecase of a customer creating their own MsalException? Nor do I see us creating a new exception, it should likely be possible to detect that by a utest if we wanted to.

---
In reply to: [264489006](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/pull/967#discussion_r264489006) [](ancestors = 264489006)

File: src/Microsoft.Identity.Client/MsalException.cs (hunk 2 above)
Comment (jmprieur) [reply src/Microsoft.Identity.Client/MsalException.cs]: I was wondering the same thing?

File: src/Microsoft.Identity.Client/MsalException.cs (hunk 3 above)
Comment (MarkZuber) [reply src/Microsoft.Identity.Client/MsalException.cs]: This is the case where we tried to deserialize a json that has a type in it we don't understand.  so we should definitely throw an exception here, shouldn't we?  thus, the todo below about adding in a real error message (i was waiting on the core error message merge to come in before i added the message properly)

File: src/Microsoft.Identity.Client/MsalException.cs (hunk 2 above)
Comment (MarkZuber) [reply src/Microsoft.Identity.Client/MsalException.cs]: The serialization isn't the hard part, it's the de-serialization.  We want to ensure the correct types can be re-hydrated on the other end and as far as I know there's not a way to do that.

**Comment by henrik-me** [comment]
As already mentioned. Only thought was why this was needed, but see no issues doing the serialization/deserialization this way.

**Pull request merged**